File: app/src/main/java/nl/viasalix/btroosterlite/TestTimetableFragment.java
/*  BTRooster Lite: Roosterapp voor Calvijn College
 *  Copyright (C) 2017 Rutger Broekhoff <rutger broekhoff three at gmail dot com>
 *                 and Jochem Broekhoff
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.viasalix.btroosterlite;

import android.app.Fragment;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity;

import java.util.ArrayList;
import java.util.List;

import static android.content.Context.CONNECTIVITY_SERVICE;

public class TestTimetableFragment extends Fragment {
    View view;
    WebView webView;
    SharedPreferences sharedPreferences;

    String code;
    String location;

    List<String> availableTestweeks = new ArrayList<>();
    List<String> availableTestweeksNames = new ArrayList<>();

    public TestTimetableFragment() {
        // Required empty public constructor
    }

    public static TestTimetableFragment newInstance() {
        TestTimetableFragment fragment = new TestTimetableFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        view = inflater.inflate(R.layout.fragment_test_timetable, container, false);
        webView = view.findViewById(R.id.web_view);
        return view;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.appbar_mainactivity_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_reload:
                loadTestTimetable(true);
                return true;
            case R.id.action_settings:
                Intent intent = new Intent(getActivity(), SettingsActivity.class);
                getActivity().startActivity(intent);
                break;
            case R.id.action_opensource:
                Intent ossIntent = new Intent(getActivity(), OssLicensesMenuActivity.class);
                ossIntent.putExtra("title", "Open-source licenties");
                getActivity().startActivity(ossIntent);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onStart() {
        super.onStart();

        Toolbar toolbar = getActivity().findViewById(R.id.toolbar);
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);

        if (((AppCompatActivity) getActivity()).getSupportActionBar() != null) {
            ((AppCompatActivity) getActivity()).getSupportActionBar().setDisplayShowTitleEnabled(false);
        }

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        code = sharedPreferences.getString("code", "12345");
        location = sharedPreferences.getString("location", "Goes");

        loadTestTimetable(true);
    }

    private boolean online() {
        ConnectivityManager connectivityManager = (ConnectivityManager) getActivity().getSystemService(CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = null;
        if (connectivityManager != null) {
            networkInfo = connectivityManager.getActiveNetworkInfo();
        }

        return (networkInfo != null && networkInfo.isConnected());
    }

    private void loadTestTimetable(boolean getIndexes) {
        webView.getSettings().setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);

        if (online()) {
            webView.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
        } else {
            webView.getSettings().setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
        }

        if (getIndexes) getIndexes();
        else            getTestTimetable();
    }

    private void getTestTimetable() {
        int weekChange = sharedPreferences.getInt("tt_weekChange", 0);

        Uri.Builder builder = new Uri.Builder();
        builder.scheme("https")
                .authority(MainActivity.AUTHORITY)
                .appendPath("ToetsroosterEmbedServlet")
                .appendQueryParameter("code", code)
                .appendQueryParameter("locatie", location)
                .appendQueryParameter("type", "leerlingen")
                .appendQueryParameter("toetsweek", availableTestweeks.get(weekChange));
        String url = builder.build().toString();

        webView.loadUrl(url);
    }

    private void getIndexes() {
        if (online()) {
            RequestQueue queue = Volley.newRequestQueue(getActivity());
            Uri.Builder builder = new Uri.Builder();
            builder.scheme("https")
                    .authority(MainActivity.AUTHORITY)
                    .appendPath("ToetsroosterEmbedServlet")
                    .appendQueryParameter("indexOphalen", "1")
                    .appendQueryParameter("locatie", location);
            String url = builder.build().toString();

            Log.d("url built", url);

            StringRequest stringRequest = new StringRequest(Request.Method.GET, url,
                    new Response.Listener<String>() {
                        @Override
                        public void onResponse(String response) {
                            sharedPreferences.edit().putString("tt_indexes", response).apply();
                            Log.d("or", response);
                            handleResponse(response);
                        }
                    }, new Response.ErrorListener() {
                @Override
                public void onErrorResponse(VolleyError error) {
                    Log.d("error", error.getMessage());
                }
            });

            queue.add(stringRequest);
        } else {
            String response = sharedPreferences.getString("tt_indexes", null);
            handleResponse(response);
            getTestTimetable();
        }
    }

    private void handleResponse(String response) {
        availableTestweeks.clear();
        availableTestweeksNames.clear();

        int i = 0;

        String[] responses = response.split("\n");

        for (String responseString : responses) {
            if (responseString.trim().length() > 0) {
                String[] responseStringSplit = responseString.split("\\|", 2);

                availableTestweeks.add(i, responseStringSplit[0]);
                availableTestweeksNames.add(i, responseStringSplit[1]);
                ++i;
            }
        }

        Spinner weekSpinner = getActivity().findViewById(R.id.tt_week_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getActivity(), android.R.layout.simple_spinner_dropdown_item, availableTestweeksNames);
        weekSpinner.setAdapter(adapter);

        weekSpinner.setSelection(sharedPreferences.getInt("tt_weekChange", 1));

        weekSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int position, long id) {
                sharedPreferences.edit().putInt("tt_weekChange", position).apply();

                loadTestTimetable(false);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {}
        });
    }
}

File: app/src/main/java/nl/viasalix/btroosterlite/TestTimetableFragment.kt
/*  BTRooster Lite: Roosterapp voor Calvijn College
 *  Copyright (C) 2017 Rutger Broekhoff <rutger broekhoff three at gmail dot com>
 *                 and Jochem Broekhoff
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.viasalix.btroosterlite

import android.app.Fragment
import android.content.Intent
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner

import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity

import java.util.ArrayList

import android.content.Context.CONNECTIVITY_SERVICE

class TestTimetableFragment : Fragment() {
    private var currentView: View? = null
    private var webView: WebView? = null
    private var weekSpinner: Spinner? = null
    private var sharedPreferences: SharedPreferences? = null

    private var code: String? = null
    private var location: String? = null

    private var availableTestweeks: MutableList<String> = ArrayList()
    private var availableTestweeksNames: MutableList<String> = ArrayList()

    private var menuHasLoaded = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        currentView = inflater.inflate(R.layout.fragment_test_timetable, container, false)
        webView = currentView!!.findViewById(R.id.web_view)
        weekSpinner = currentView!!.findViewById(R.id.tt_week_spinner)

        return currentView
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.appbar_mainactivity_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_reload -> {
                loadTestTimetable(true)
                return true
            }
            R.id.action_settings -> {
                val intent = Intent(activity, SettingsActivity::class.java)
                activity.startActivity(intent)
            }
            R.id.action_opensource -> {
                val ossIntent = Intent(activity, OssLicensesMenuActivity::class.java)
                ossIntent.putExtra("title", "Open-source licenties")
                activity.startActivity(ossIntent)
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onStart() {
        super.onStart()

        val toolbar = activity.findViewById<Toolbar>(R.id.toolbar)
        (activity as AppCompatActivity).setSupportActionBar(toolbar)

        if ((activity as AppCompatActivity).supportActionBar != null)
            (activity as AppCompatActivity)
                    .supportActionBar!!
                    .setDisplayShowTitleEnabled(false)

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)
        code = sharedPreferences!!.getString("code", "12345")
        location = sharedPreferences!!.getString("location", "Goes")

        loadTestTimetable(true)
    }

    private fun online(): Boolean {
        val connectivityManager = activity.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo: NetworkInfo?

        networkInfo = connectivityManager.activeNetworkInfo

        return networkInfo != null && networkInfo.isConnected
    }

    private fun loadTestTimetable(getIndexes: Boolean) {
        webView!!.settings.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK

        if (online())
            webView!!.settings.cacheMode = WebSettings.LOAD_NO_CACHE
        else
            webView!!.settings.cacheMode = WebSettings.LOAD_CACHE_ELSE_NETWORK

        if (getIndexes)
            getIndexes()
        else
            getTestTimetable()
    }

    private fun getTestTimetable() {
        val weekChange = sharedPreferences!!.getInt("tt_weekChange", 0)

        val builder = Uri.Builder()
        builder.scheme("https")
                .authority(MainActivity.AUTHORITY)
                .appendPath("ToetsroosterEmbedServlet")
                .appendQueryParameter("code", code)
                .appendQueryParameter("locatie", location)
                .appendQueryParameter("type", "leerlingen")
                .appendQueryParameter("toetsweek", availableTestweeks[weekChange])
        val url = builder.build().toString()

        webView!!.loadUrl(url)
    }

    private fun getIndexes() {
        if (online()) {
            val queue = Volley.newRequestQueue(activity)
            val builder = Uri.Builder()
            builder.scheme("https")
                    .authority(MainActivity.AUTHORITY)
                    .appendPath("ToetsroosterEmbedServlet")
                    .appendQueryParameter("indexOphalen", "1")
                    .appendQueryParameter("locatie", location)
            val url = builder.build().toString()

            val stringRequest = StringRequest(Request.Method.GET, url,
                    { response ->
                        sharedPreferences!!.edit().putString("tt_indexes", response).apply()
                        Log.d("or", response)
                        handleResponse(response)
                    }) { error -> Log.d("error", error.message) }

            queue.add(stringRequest)
        } else {
            val response = sharedPreferences!!.getString("tt_indexes", null)
            handleResponse(response)
            getTestTimetable()
        }
    }

    private fun handleResponse(response: String?) {
        availableTestweeks.clear()
        availableTestweeksNames.clear()

        var i = 0

        val responses = response!!.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        for (responseString in responses) {
            if (responseString.isNotEmpty()) {
                val responseStringSplit = responseString.split("\\|".toRegex(), 2).toTypedArray()

                availableTestweeks.add(i, responseStringSplit[0])
                availableTestweeksNames.add(i, responseStringSplit[1])
                ++i
            }
        }

        val adapter = ArrayAdapter(
                activity,
                android.R.layout.simple_spinner_dropdown_item,
                availableTestweeksNames)

        weekSpinner!!.adapter = adapter

        menuHasLoaded = true

        weekSpinner!!.setSelection(sharedPreferences!!.getInt("tt_weekChange", 1))

        weekSpinner!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, position: Int, id: Long) {
                sharedPreferences!!.edit().putInt("tt_weekChange", position).apply()

                loadTestTimetable(false)
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }
    }
}// Required empty public constructor
